The RF ModCall module takes two RC files generated by the RF Count module, and performs transcriptome-wide single-base resolution calling of &Psi;/2'-OMe residues from &Psi;-seq/Pseudo-seq and 2OMe-seq experiments.

!!! note "Information"
    For more details, please refer to Carlile *et al*., 2014 (Pseudo-seq, PMID: [25192136](https://www.ncbi.nlm.nih.gov/pubmed/25192136)), Schwartz *et al*., 2014 (&Psi;-seq, PMID: [25219674](https://www.ncbi.nlm.nih.gov/pubmed/25219674)), and Incarnato *et al*., 2016 (2OMe-seq, PMID: [27614074](https://www.ncbi.nlm.nih.gov/pubmed/27614074)).
For each transcript's position, two measures are computed:<br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mi>w</mi><mo>&#xD7;</mo><mfrac><mrow><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><mo>-</mo><msub><mi>n</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub></mrow><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mi>i</mi><mo>-</mo><mstyle displaystyle="true"><mfrac bevelled="true"><mi>w</mi><mn>2</mn></mfrac></mstyle></mrow><mrow><mi>i</mi><mo>+</mo><mstyle displaystyle="true"><mfrac bevelled="true"><mi>w</mi><mn>2</mn></mfrac></mstyle></mrow></munderover><mo>(</mo><msub><mi>n</mi><mrow><mi>T</mi><mi>j</mi></mrow></msub><mo>+</mo><msub><mi>n</mi><mrow><mi>U</mi><mi>j</mi></mrow></msub><mo>)</mo><mo>&#xA0;</mo><mo>-</mo><mo>&#xA0;</mo><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><mo>-</mo><msub><mi>n</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub></mrow></mfrac></math>
<br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>R</mi><mi>i</mi></msub><mo>=</mo><mfrac><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub></mfrac></math>
<br/>
where *S<sub>i</sub>* and *R<sub>i</sub>* are respectively the score and the ratio at position *i* of the transcript, *w* is the size (in nt) of a window centered on position *i*, *n<sub>Ti</sub>* and *n<sub>Ui</sub>* are respectively the number of RT-stops in the CMCT treated (or low dNTP) and CMCT untreated (or high dNTP) samples, and *c<sub>Ti</sub>* is the read coverage at position *i* in the CMCT treated (or low dNTP) sample.<br/>
The __score__ is a measure of the RT-stop enrichment in the CMCT treated (or low dNTP) sample at a given position, with respect to the surrounding bases, and to the CMCT untreated (or high dNTP) sample.<br/>
The __ratio__ is a relative quantitation of the modification stoichiometry at a given position in the CMCT treated (or low dNTP) sample.<br/>

!!! warning "Warning"
    When processing SAM/BAM files from &Psi;-seq/Pseudo-seq or 2OMe-seq experiments with RF Count, __avoid using__ the ``--no-mapped-count`` parameter, otherwise RF ModCall will not be able to perform library size scaling.<br/>
    
# Usage
To list the required parameters, simply type:

```bash
$ rf-modcall -h
```

Parameter         | Type | Description
----------------: | :--: |:------------
__-u__ *or* __--untreated__ | string | Path to the RC file for the CMCT untreated (or high dNTP)
__-t__ *or* __--treated__ | string | Path to the RC file for the CMCT treated (or low dNTP) sample
__-i__ *or* __--index__ | string[,string] | A comma separated (no spaces) list of RCI index files for the provided RC files<br/>__Note #1:__ RCI files must be provided in the order 1. Untreated, 2. Treated<br/>__Note #2:__ If a single RTI file is specified, it will be used for all RC files<br/>__Note #3:__ If no RCI index is provided, it will be generated at runtime, and stored in the same folder of the untreated/treated samples
__-p__ *or* __--processors__ | int | Number of processors (threads) to use (Default: __1__)
__-o__ *or* __--output-dir__ | string | Output directory for writing site scores and ratios in XML format (Default: __&lt;treated&gt;\_vs\_&lt;untreated&gt;/__)
__-ow__ *or* __--overwrite__ | | Overwrites the output directory if already exists
__-w__ *or* __--window__ | int | Window size (in nt) for score calculation (&ge;3, Default: __150__)
__-ts__ *or* __--to-smaller__ | | The larger sample will be scaled toward the smaller one (Default: __scale smaller sample to the larger one__)
__-mc__ *or* __--mean-coverage__ | float | Discards any transcript with mean coverage below this threshold (&ge;0, Default: __0__)
__-ec__ *or* __--median-coverage__ | float | Discards any transcript with median coverage below this threshold (&ge;0, Default: __0__)
__-D__ *or* __--decimals__ | int | Number of decimals for reporting scores/ratios (1-10, Default: __3__)
__-n__ *or* __--nan__ | int | Positions of transcript with read coverage behind this threshold, will be reported as NaN in the reactivity profile (&gt;0, Default: __10__)

<br/>
## Output XML files
RF ModCall produces a XML file for each transcript being analyzed, with the following structure:<br/>

```xml
<?xml version="1.0" encoding="UTF-8"?><data [attributes]>	<transcript id=”Transcript ID” length=”Transcript length”>		<sequence>			Transcript sequence		</sequence>		<score>			Comma-separated list of scores		</score>
		<ratio>			Comma-separated list of ratios		</ratio>	</transcript></data>
```
The __data__ tag’s attributes allow keeping track of the analysis performed:<br/>

Attribute     | Possible values | Description
-------------: | :------------: | :----------
__win__ | Positive integer &ge; 3 | Window's size (in nt) for score calculation
__tosmaller__ | TRUE/FALSE | Whether the larger dataset has been scaled to the size of the smaller one